### ISSUE TITLE
Add underlying API return code and error message to better support terraform check scenario.

### ISSUE THREAD

**Issue by LaurentLesle** [magodo/terraform-provider-restful#66]
With the new support of the Terraform check block , the restful provider would provide a very nice way to assert some condition.
The goal is to write integration tests like triggering an operation that is expected to work or fail. Trying to access a resource that is not authorised....

This issue is to request the addition of a return value of the underlying API call (like 200, 403...)

Here an example:

```hcl
check "sp_x_should_not_have_read_access_to_tenant_root" {
  data "restful_resource" "management_group_tenant_root" {
    id = format("providers/Microsoft.Management/managementGroups/%s", local.running_context.tenant_id)
    query = {
      api-version = ["2020-05-01"]
    }
  }

  assert {
    condition     = data.restful_resource.management_group_tenant_root.api_return_code == 403
    error_message = "Service principal x can read the Management Group Tenant root."
  }
}

```

Proposition:

Add attributes:

- api_return_code
- api_error_message

to:

- data.restful_resource
- restful_resource
- restful_operation

**Comment by magodo** [comment]
@LaurentLesle I feel these kinds of details only applies to `restful_operation`? So we probably want a new property to indicate tolerating any API error?

**Comment by LaurentLesle** [comment]
I think you are correct we need an attribute to tolerate any API error.

The reason I am asking to cover the 2 resources and the data source is to cover some scenarios like:

"Try to create a resource group in an environment you are not authorised to. If is failing then it is a positive test case result."
"Try to create a resource group in an authorised environment. If return code is 200 then it is a positive test case result"

Sometimes the tests would require to read (success) or try to read an unauthorized object (access denied)
Sometimes to trigger an operation on an object
Sometimes to create / delete an object,

The goal is to leverage this feature in terraform test and check block.

**Comment by magodo** [comment]
@LaurentLesle Per the document of [`check`](https://developer.hashicorp.com/terraform/language/checks):

> You can declare a check block with a local name, zero-to-one scoped [data sources](https://developer.hashicorp.com/terraform/language/checks#scoped-data-sources), and one-to-many [assertions](https://developer.hashicorp.com/terraform/language/checks#assertions).

Only data source is accepted in the `check` block. The scenario you mentioned here seems not quite fit to the philosophy of terraform given that a successfully provisioned resource represents an existing resource. The problem of introducing such a flag to a resource is that what is then expected to happen when users update/read/delete this resource.

Also, for supporting methods other than `GET` to the `restful_resource` data source, it might be a valid case to extend to `HEAD`/`POST`, as https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/http#method. But we definitely don't want to extend it to support `PUT`.

**Comment by LaurentLesle** [comment]
Agreed for check block. That would make sense to add also HEAD/POST in the data.restful_resource , similar to the http data source and it can also be used in precondition and postcondition.

What is also your view on adding the api_response_code to resource restful_operation?

**Comment by magodo** [comment]
@LaurentLesle For the `restful_operation` resource, since it's `read` and `delete` is noop, so it should be fine to add the api error toleration flag to it. Whilst I'm wondering whether there is a valid scenario for it to be useful?
